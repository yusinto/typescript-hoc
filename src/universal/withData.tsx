import * as React from 'react';

export interface Post {
  userId: number;
  id: number;
  title: string;
  body: string;
}

export interface InjectedProps {
  data: any | Post;
}

interface HocState extends InjectedProps {
  isLoading: boolean;
}

export default function withData<P>(WrappedComponent: React.ComponentType<P & InjectedProps>, dataUrl: string) {
  return class extends React.Component<P, HocState> {
    readonly state: HocState = { isLoading: true, data: {} };

    fetchData = async () => {
      const response = await fetch(dataUrl);
      const data = await response.json();
      this.setState({ isLoading: false, data });
    };

    async componentDidMount() {
      await this.fetchData();
    }

    render() {
      const { isLoading, data } = this.state;
      return isLoading ? 'Loading' : <WrappedComponent {...this.props} data={data} />;
    }
  };
}
