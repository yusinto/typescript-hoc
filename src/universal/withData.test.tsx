import 'jest-fetch-mock';
import * as React from 'react';
import { create } from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';
import withData, { EnhancedComponent, InjectedProps, Post } from './withData';

interface Data {
  title?: string;
  body?: string;
}

interface DataRoot {
  data?: Data;
}
const { fetch } = global;
const mockFetchUrl = 'https://api.yus.com/post/1';
const App = (props: DataRoot) => (
  <div>
    {props.data && props.data.title}
    <br />
    {props.data && props.data.body}
  </div>
);
const mockResponse = {
  userId: 6,
  id: 5,
  title: 'Test title',
  body: 'This is a test body',
} as Post;

describe('withData', () => {
  beforeEach(() => {
    fetch.resetMocks();
    fetch.mockResponse(JSON.stringify(mockResponse));
  });

  test('renders loading initially', async () => {
    const Enhanced = withData(App, mockFetchUrl);
    const component = create(<Enhanced />);
    expect(component).toMatchSnapshot();
  });

  test.only('componentDidMount fetches data', async () => {
    const Enhanced = withData(App, mockFetchUrl);
    const component = mount(<Enhanced />);
    const instance = component.instance() as EnhancedComponent;
    await instance.componentDidMount();

    component.update();
    // expect(fetch.mock.calls.length).toEqual(1);
    // expect(fetch.mock.calls[0][0]).toEqual(mockFetchUrl);
    expect(component).toMatchSnapshot();
  });

  test('fetchData', () => {});

  test('render loading', () => {});

  test('render component', () => {});
});
