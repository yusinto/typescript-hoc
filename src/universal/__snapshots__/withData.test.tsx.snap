// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withData componentDidMount fetches data 1`] = `
<class_1>
  <App
    data={
      Object {
        "body": "This is a test body",
        "id": 5,
        "title": "Test title",
        "userId": 6,
      }
    }
  >
    <div>
      Test title
      <br />
      This is a test body
    </div>
  </App>
</class_1>
`;

exports[`withData renders loading initially 1`] = `"Loading"`;
